### DISCUSSION TITLE
Simple API for manual nesting?

### DISCUSSION THREAD

**Discussion by jrochkind** [okuramasafumi/alba#233] [Ideas]
I am working with alba to see if I can easily do what I need with it. 

I had an idea, what if alba could have simple API for doing "manual" nesting?

Imagine Employee with `name`, `address_street`, `address_country`. 

```ruby
class EmployeeResource
  include Alba::Resource
 
  attributes :name
 
  nested :address
     attribute :street, &:address_street
     attribute  :country, &:address_country
  end
end
```

To result in:

```ruby
{
  name: "some name"
  address: {
    street: "some street",
    country: "some country"
  }
}
```

`nested`  would just create a nested hash with that name -- inside the block, you are still in the same context as ever, and can call `attributes` and `attribute` that just work as usual on the "top" object. 

(I am not sure `nested` is exactly the right name for this method). 

This would get at the use case in https://github.com/okuramasafumi/alba/discussions/211.  There you suggested using Rails `composed_of`, but I don't want to add functionality to my model that is really related to the "shape" I want to serialize in a specific context. 

There you also found another way using addiitonal "resource" classes. This is possible, but can get inconvenient when you have a lot of nesting. 

For instance, this feature maybe also could be used for implementing Json:Api actually -- in a flexible way that lets someone build Json:Api _or_ something similar that isn't exactly Json:Api , fairly simply:

```ruby
class EmployeeResource
    include Alba::Resource

   nested :data do
      attribute :type do
         "article"
      end
      attribute :id 

      nested :attributes do
         attributes :foo, :baz. 

         # I actually want to treat this as an attribute, which is totally legal in
         # json:api, an attribute whose value is complex json. This is something
         # your existing json:api draft I don't think lets me easily do
        many :departments do
             attributes :name, :type
         end
      end

      nested :relationships do
         many :something do
             nested :data do
                attributes :id 
                attribute: type do
                    "something"
                end 
         end
      end  
 
      nested :links do
          attribute :self, &:something
      end
   end
end
```
```
{
    data: {
       type: "article",
       id: "1",
       attributes: {
          # ....
       },
      relationships: {
       # etc
```

I don't think this will get all the way to json:api -- For instance, not sure how to use it to handle json:api _collections_, and relationships are still too fiddly.  But maybe it is helpful in defining json:api in simpler terms?  

And it lets you do a variety of flexible things that are just similar to json:api in some ways. A simple additional alba API that makes easier a whole category of things you might want to do to "shape" your output. 

This is just an idea right now for discussion, I'm not completely positive if it's a great idea.

**Comment by okuramasafumi** [comment]
Thank you for the suggestion. Basically I'd like the "model" layer to expose nested data structure (in this example, `address`), but yes, as you said it's flexible, and if it's optional feature we won't feel like we'd like to remove it anyway.
One thing we need to consider is a name, `nested` is too abstract, but other names, such as `hash`, don't sound right.
This nested thing is called `object` in JSON spec (ref: https://www.crockford.com/mckeeman.html) but `object` is already taken for underlying object.

**Reply by jrochkind** [reply]
Thanks for your attention!

> Basically I'd like the "model" layer to expose nested data structure 

I honestly feel the opposite. That there needs to be a "nesting" is a property of the serialization, not of the model itself. Different serializations for different consumers may need different "shapes".  I don't want to put it in my model when it's really determined by what I want the serialization to look like.  

I want to be able to have flexible serializations of any model, where the serialization can change without needing to change the model to support the new serialization. 

I suppose I could try to create some kind of intermediate "model" that isn't really my base model, is like some kind of proxy object that changes things to provide the nested structures for the serialiazations... but i don't want to, that's going to be some confusing fragile code, in my experience. 

I don't find `nested` to be abstract personally, but I'm open to any naming.  How did y'all refer to it in #211? The title there is "nested".  You also say "Alba currently doesn't have an API to define nested hash easily," using the word "nested" there too. I suppose `nested_hash` would be a very explicit alternative. `compound` to me seems more confusing.

**Reply by okuramasafumi** [reply]
I'm convinced that serializers are responsible for nesting and creating proxy object for this purpose is not a good idea. Let's add this feature to Alba.

I'd prefer `nested_hash` because `nested` is not clear about what it returns.

**Reply by trevorturk** [reply]
I think this looks like a great feature! I just wanted to chime in about the name, I think perhaps `nested_hash` is a bit odd, since you can return a `Hash` or a `JSON` object? I think `nested` could be fine. I looked for prior art and found this: https://github.com/procore/blueprinter/issues/239 -- to compare, I think Alba would do something like:

```ruby
class EmployeeResource
  include Alba::Resource

  attribute :address do
     attribute :street, &:address_street
  end
end
```

I could also see using something like `nested_attributes` or perhaps the word `wrapped` could be considered. I noticed that Rails uses that term a bit, for example in: https://api.rubyonrails.org/v7.0.3/classes/ActionController/ParamsWrapper.html

Anyway, please feel free to ignore me. This looks like a lovely feature, I just couldn't resist sharing my bike-shedding comments, sorry! 😬

**Reply by okuramasafumi** [reply]
I also came up with the idea that we can simply call `attribute` within `attribute`, in other words, `attribute` is simply nestable.

If that's difficult we could pass `nested: true` option to `attribute` call, that solves naming issue.

**Reply by trevorturk** [reply]
This looks great, and I think the implementation in #237 seems pretty light weight, which is very nice. I think you'll find the right path within these good options. Excellent work, you two! Thank you! 💝

**Reply by jrochkind** [reply]
Simply calling `attribute` within `attribute` is a nice API, I agree, but since a block argument to `attribute` already exists with another purpose, I think the implementation might be complicated, to kind of let it do two things at once. Might not be worth it.
